app/utils/graph-area.js

```diff
@@ -0,0 +1,240 @@
+    .append('rect').attr('width', width).attr('height', height);
+
+  const stripes = drawStripes(svg, width, y, yTicks);
+
+  const series = getSeries(fields, svg, gradient);
+
+  const tooltip = initTooltip({ el, svg, height, margin, maxPoints, duration, formatter: FORMATTERS[formatter], x, gradient: GRADIENT_COLORS[gradient] });
+
+  const params = {
+    x,
+    y,
+    min,
+    height,
+    line,
+    area,
+    yAxisG,
+    yAxis,
+    options,
+    maxPoints,
+    query,
+    series,
+    duration,
+    tooltip,
+    stopped: false,
+  }
+
+  return {
+    start() {
+      start(params);
+    },
+    fit() {
+      fit({ el, x, margin, graph, clipPath, stripes });
+    },
+    destory() {
+      params.stopped = true;
+    }
+  };
+}
+
+function start(params) {
+  if (params.stopped) {
+    return
+  }
+
+  let all = [];
+  const series = [];
+  params.series.forEach(serie => {
+    const data = query(params.maxPoints, params.query, serie.field);
+    all = all.concat(data)
+    series.push({
+      field: serie.field,
+      lineChart: serie.lineChart,
+      areaChart: serie.areaChart,
+      data
+    })
+  })
+
+  updateTooltip(series, params)
+  updateAxis(all, params)
+  updateLines(series, params)
+}
+
+function fit(params) {
+  const margin = params.margin;
+  const width = getWeight(params.el, margin);
+
+  params.x.range([0, width]);
+  params.graph.attr('width', width + margin.left + margin.right);
+  params.clipPath.attr('width', width);
+  params.stripes.attr('width', width);
+}
+
+function query(maxPoints, query, field) {
+  let data = query(field.key) || []
+
+  if (data.length < maxPoints) {
+    const len = data.length
+    for (let i = 0; i < maxPoints - len; i++) {
+      data.unshift(null)
+    }
+  } else if (data.length > maxPoints) {
+    data = data.slice(-1 * maxPoints);
+  }
+
+  return data;
+}
+
+function adjustMax(dataMax, options) {
+  let optMinMax = options.minMax;
+  let optMax = options.max;
+  let optScaleDown = options.scaleDown;
+  let observedMax = options.observedMax;
+  let out = dataMax;
+
+  if (optMax) {
+    out = optMax;
+  } else if (optMinMax) {
+    out = Math.max(optMinMax, out);
+  }
+
+  if (observedMax && !optScaleDown) {
+    out = Math.max(observedMax, out);
+  }
+
+  if (!observedMax && out > 0 && options.maxDoubleInital) {
+    out *= 2;
+  }
+  options.observedMax = out;
+  return out;
+}
+
+function updateTooltip(series, params) {
+  params.tooltip.update(series)
+}
+
+function updateAxis(all, params) {
+  const min = params.min === null ? d3.min(all) : params.min;
+  const max = adjustMax(d3.max(all), params.options);
+
+  params.y.domain([min, max]);
+  params.y.range([params.height - 2, 2]);
+  params.y.rangeRound([params.height - 2, 2]);
+  params.yAxisG.call(params.yAxis);
+}
+
+function updateLines(series, params) {
+  series.forEach((serie, i) => {
+
+    serie.areaChart.attr('d', params.area(serie.data))
+      .attr('transform', null)
+      .transition()
```

Comment:
The smooth animation looks nice but is very cpu-intensive; can you make it update only when the data changes?